Flowblade

polygon opened the pull request:
Multitrack Non-Linear Video Editor

Seems to run well. I created a simple project with some clips, done some editing, and rendered it out. All of that worked flawlessly.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

polygon committed:
mlt: Add option to build python bindings

Python bindings are required for flowblade package
Also updated licenses according to copying information

polygon committed:
flowblade: init at 2.8.0.3

polygon commented:
nixpkgs-review result

4 packages built:
flowblade mlt python38Packages.mlt-bindings python39Packages.mlt-bindings

r-rmcgibbo commented:
Result of `nixpkgs-review pr 126569` at 3571fb2d run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built successfully:</summary>
  <ul>
    <li>flowblade</li>
    <li>mlt</li>
    <li>python38Packages.mlt-bindings</li>
    <li>python39Packages.mlt-bindings</li>
  </ul>
</details>
<details>
  <summary>6 suggestions:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `pkgs`.
Near pkgs/development/python-modules/mlt-bindings/default.nix:1:8:
```
  |
1 | { lib, pkgs, mlt, python, buildPythonPackage, pythonOlder }:
  |        ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl3` is a deprecated license, please check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/applications/video/flowblade/default.nix:45:5:
```
   |
45 |     license = licenses.gpl3;
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

pkg-config is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/libraries/mlt/default.nix:19:3:
```
   |
19 |   buildInputs = [
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

makeWrapper is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/libraries/mlt/default.nix:19:3:
```
   |
19 |   buildInputs = [
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl3` is a deprecated license, please check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/development/libraries/mlt/default.nix:75:5:
```
   |
75 |     license = licenses.gpl3;
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl3` is a deprecated license, please check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/development/python-modules/mlt-bindings/default.nix:18:5:
```
   |
18 |     license = licenses.gpl3;
   |     ^
```
</li>
  </ul>
</details>

polygon commented:
Updated according to auto-review suggestions

davidak commented:
Result of `nixpkgs-review pr 126569` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>flowblade</li>
    <li>mlt</li>
    <li>python38Packages.mlt-bindings</li>
    <li>python39Packages.mlt-bindings</li>
  </ul>
</details>

davidak commented on pkgs/applications/video/flowblade/default.nix:
```diff
@@ -0,0 +1,49 @@
+
+  meta = with lib; {
+    description = "Multitrack Non-Linear Video Editor";
+    homepage = "https://github.com/jliljebl/flowblade";
```
```suggestion
    homepage = "http://jliljebl.github.io/flowblade/";
```

polygon replied:
Agreed

davidak commented on pkgs/applications/video/flowblade/default.nix:
```diff
@@ -0,0 +1,49 @@
+  meta = with lib; {
+    description = "Multitrack Non-Linear Video Editor";
+    homepage = "https://github.com/jliljebl/flowblade";
+    license = with licenses; [ gpl3Plus gpl2Plus ];
```
```suggestion
    license = with licenses; [ gpl3Plus ];
```

gpl2Plus are only some files. so we can use them as gpl3Plus and that is the license of the complete package

https://github.com/jliljebl/flowblade/blob/master/flowblade-trunk/copyrights

polygon replied:
Agreed

davidak commented:
I noticed that the preview is transparent. The rendered video looks fine.

I use the Pantheon desktop (GTK based). Maybe it works fine on other desktops.

https://youtu.be/R-S0wkmb0h8

Except that it seem to work fine. It didn't crash a single time!

davidak commented:
There are also errors that filters and transitions are not found. Is that expected or can it be fixed?



<details>
  <summary>Console output</summary>
  
  ```
Loading render profiles...
Loading filters...
MLT service frei0r.alphagrad not found.
MLT service frei0r.mask0mate not found.
MLT service frei0r.alphaspot not found.
MLT service frei0r.alpha0ps not found.
MLT service frei0r.select0r not found.
MLT service frei0r.spillsupress not found.
MLT service ladspa.1406 not found.
MLT service ladspa.1433 not found.
MLT service ladspa.1408 not found.
MLT service ladspa.1431 not found.
MLT service ladspa.1901 not found.
MLT service ladspa.1438 not found.
MLT service ladspa.1195 not found.
MLT service ladspa.1402 not found.
MLT service ladspa.1185 not found.
MLT service ladspa.1213 not found.
MLT service ladspa.1904 not found.
MLT service ladspa.1903 not found.
MLT service ladspa.1215 not found.
MLT service ladspa.1216 not found.
MLT service ladspa.1410 not found.
MLT service ladspa.1892 not found.
MLT service ladspa.1194 not found.
MLT service ladspa.1197 not found.
MLT service ladspa.1423 not found.
MLT service ladspa.1910 not found.
MLT service ladspa.1417 not found.
MLT service ladspa.1210 not found.
MLT service ladspa.1198 not found.
MLT service ladspa.1905 not found.
MLT service ladspa.1201 not found.
MLT service frei0r.glow not found.
MLT service frei0r.scanline0r not found.
MLT service frei0r.cartoon not found.
MLT service frei0r.vignette not found.
MLT service frei0r.emboss not found.
MLT service frei0r.dither not found.
MLT service frei0r.colorhalftone not found.
MLT service frei0r.softglow not found.
MLT service frei0r.pixeliz0r not found.
MLT service frei0r.squareblur not found.
MLT service frei0r.IIRblur not found.
MLT service frei0r.rgbnoise not found.
MLT service frei0r.contrast0r not found.
MLT service frei0r.saturat0r not found.
MLT service frei0r.hueshift0r not found.
MLT service frei0r.levels not found.
MLT service frei0r.coloradj_RGB not found.
MLT service frei0r.coloradj_RGB not found.
MLT service frei0r.colgate not found.
MLT service frei0r.cairogradient not found.
MLT service frei0r.cluster not found.
MLT service frei0r.threelay0r not found.
MLT service frei0r.threshold0r not found.
MLT service frei0r.primaries not found.
MLT service frei0r.colordistance not found.
MLT service frei0r.posterize not found.
MLT service frei0r.colortap not found.
MLT service frei0r.sigmoidaltransfer not found.
MLT service frei0r.tint0r not found.
MLT service frei0r.colorize not found.
MLT service frei0r.distort0r not found.
MLT service frei0r.defish0r not found.
MLT service frei0r.lenscorrection not found.
MLT service frei0r.nosync0r not found.
MLT service frei0r.cairoimagegrid not found.
MLT service frei0r.edgeglow not found.
MLT service frei0r.sobel not found.
MLT service frei0r.hqdn3d not found.
MLT service frei0r.sharpness not found.
MLT service frei0r.letterb0xed not found.
MLT service frei0r.baltan not found.
MLT service frei0r.vertigo not found.
MLT service frei0r.nervous not found.
MLT service frei0r.mask0mate not found.
Loading transitions...
MLT transition frei0r.cairoaffineblend not found.
MLT transition frei0r.cairoblend not found.
MLT transition frei0r.darken not found.
MLT transition frei0r.hardlight not found.
MLT transition frei0r.addition not found.
MLT transition frei0r.burn not found.
MLT transition frei0r.difference not found.
MLT transition frei0r.divide not found.
MLT transition frei0r.color_only not found.
MLT transition frei0r.dodge not found.
MLT transition frei0r.grain_extract not found.
MLT transition frei0r.grain_merge not found.
MLT transition frei0r.hue not found.
MLT transition frei0r.lighten not found.
MLT transition frei0r.multiply not found.
MLT transition frei0r.overlay not found.
MLT transition frei0r.saturation not found.
MLT transition frei0r.screen not found.
MLT transition frei0r.softlight not found.
MLT transition frei0r.subtract not found.
MLT transition frei0r.value not found.
MLT transition frei0r.alphaxor not found.
MLT transition frei0r.alphain not found.
MLT transition frei0r.alphaout not found.
  ```
  
</details>

polygon commented:
Many thanks for the thorough review and testing, much appreciated.

I am not able to reproduce the issue with the transparent preview window. I've found this upstream issue that looks very similar to what you are seeing, but there does not seem to be a definitive solution:

https://github.com/jliljebl/flowblade/issues/619

For the missing filters, there seems to be a path issue. strace shows the following:

```
openat(AT_FDCWD, "/home/jan/.frei0r-1/lib", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/local/lib/frei0r-1", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/opt/local/lib/frei0r-1", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib64/frei0r-1", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/frei0r-1", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
```

This seems to be searching for the folder with all the frei0r shared libraries through some mechanism that the automatic patching does not cover. I'll be digging a bit into this.

polygon commented:
Was able to fix the frei0r-plugin errors through setting `FREI0R_PATH` in the wrapper. Some errors remain about `ladspa` plugins which are not found which I am still tracing.

polygon committed:
mlt: Add jackrack and ladspa support

polygon commented:
Fixed frei0r plugin detection and added jackrack/ladspa support to mlt so that ladspa plugins are now also supported. Also added `ladspaPlugins` dependency to flowblade so that the plugins are actually found.

SuperSandro2000 commented on maintainers/maintainer-list.nix:
```diff
@@ -8155,6 +8155,12 @@
     githubId = 1829032;
     name = "Paul Hendry";
   };
+  polygon = {
```
Please put this into a separate commit with the message ``maintainers: add NAME`` and sort it alphabetically.

polygon replied:
done

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -4344,6 +4344,10 @@ in {
   mlxtend = callPackage ../development/python-modules/mlxtend { };
 
+  mlt-bindings = callPackage ../development/python-modules/mlt-bindings {
+    mlt = (callPackage ../development/libraries/mlt { python = self.python; enablePython = true; });
```
Should we move this to top-level.nix?

polygon replied:
I've refactored the Python3 MLT bindings thanks to the answer of jtojnar here:

https://discourse.nixos.org/t/libxml-not-found-with-dependent-build/13603/5

The `development/python-modules/mlt-bindings` package definiton was removed altogether.

SuperSandro2000 commented on pkgs/development/python-modules/mlt-bindings/default.nix:
```diff
@@ -0,0 +1,22 @@
+
+  nativeBuildInputs = [ mlt ];
+
+  doCheck = false;
```
Please add a comment why you disabled tests.

polygon replied:
Removed completely after refactoring

SuperSandro2000 commented on pkgs/development/libraries/mlt/default.nix:
```diff
@@ -16,18 +18,24 @@ stdenv.mkDerivation rec {
   configureFlags = [
     "--avformat-swscale" "--enable-gpl" "--enable-gpl3" "--enable-opengl"
-  ];
+  ]
+  ++ lib.optional enablePython "--swig-languages=python";
```
```suggestion
  ] ++ lib.optional enablePython "--swig-languages=python";
```

SuperSandro2000 commented on pkgs/development/libraries/mlt/default.nix:
```diff
@@ -16,18 +18,24 @@ stdenv.mkDerivation rec {
+    SDL ffmpeg frei0r libjack2 libdv libsamplerate libvorbis libxml2.dev
+    movit sox libexif gtk2 fftw libebur128 opencv4 SDL2 jack2
+    ladspa-sdk
+  ]
+  ++ lib.optional enablePython ncurses;
```
```suggestion
  ] ++ lib.optional enablePython ncurses;
```

SuperSandro2000 commented on pkgs/development/libraries/mlt/default.nix:
```diff
@@ -2,6 +2,8 @@
 , libvorbis, libxml2, movit, pkg-config, sox, fftw, opencv4, SDL2
 , gtk2, genericUpdater, common-updater-scripts, libebur128
+, jack2, ladspa-sdk, swig, which, ncurses
+, enablePython ? false, python ? null
```
```suggestion
, enablePython ? false, python
```

SuperSandro2000 commented on pkgs/applications/video/flowblade/default.nix:
```diff
@@ -0,0 +1,51 @@
+, gobject-introspection, makeWrapper, wrapGAppsHook
+}:
+
+let
+  python_env = python3.withPackages (ps: with ps; [
+    mlt-bindings
+    pygobject3
+    dbus-python
+    numpy
+    pillow
+  ]);
+in
+stdenv.mkDerivation rec {
```
```suggestion
stdenv.mkDerivation rec {
```

SuperSandro2000 commented on pkgs/applications/video/flowblade/default.nix:
```diff
@@ -0,0 +1,51 @@
+    sha256 = "sha256-/EkI3qiceB5eKTVQnpG+z4e6yaE9hDtn6I+iN/J+h/g=";
+  };
+
+  buildInputs = [ ffmpeg frei0r sox gtk3 python_env gobject-introspection ladspaPlugins ];
```
```suggestion
  buildInputs = [ ffmpeg frei0r sox gtk3 gobject-introspection ladspaPlugins ]
    ++ python3.withPackages (ps: with ps; [ mlt-bindings pygobject3 dbus-python numpy pillow ]);
```

polygon replied:
Got an `error: value is a set while a list was expected` here. Added it inside the list directly. Hope that works, too,

polygon committed:
maintainers: add polygon

polygon commented:
Thanks for the review, Sando!

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 126569` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
4 packages built:
</summary>
<ul>
<li>flowblade</li>
<li>mlt</li>
<li>python38Packages.mlt</li>
<li>python39Packages.mlt</li>
</ul>
</details>
